File: clients/rust/crates/rubin-consensus/src/pow.rs
use crate::constants::{POW_LIMIT, TARGET_BLOCK_INTERVAL, WINDOW_SIZE};
use crate::error::{ErrorCode, TxError};
use crate::{block_hash, BLOCK_HEADER_BYTES};
use num_bigint::BigUint;
use num_traits::{One, Zero};

pub fn retarget_v1(
    target_old: [u8; 32],
    timestamp_first: u64,
    timestamp_last: u64,
) -> Result<[u8; 32], TxError> {
    let pow_limit = BigUint::from_bytes_be(&POW_LIMIT);
    let t_old = BigUint::from_bytes_be(&target_old);
    if t_old.is_zero() {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "retarget: target_old is zero",
        ));
    }
    if t_old > pow_limit {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "retarget: target_old above pow_limit",
        ));
    }

    let t_actual = if timestamp_last <= timestamp_first {
        1u64
    } else {
        timestamp_last - timestamp_first
    };

    let t_expected = TARGET_BLOCK_INTERVAL
        .checked_mul(WINDOW_SIZE)
        .ok_or_else(|| TxError::new(ErrorCode::TxErrParse, "retarget: t_expected overflow"))?;
    if t_expected == 0 {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "retarget: t_expected is zero",
        ));
    }

    // floor(target_old * T_actual / T_expected)
    let mut t_new = (&t_old * BigUint::from(t_actual)) / BigUint::from(t_expected);

    // clamp lower = max(1, floor(target_old / 4))
    let mut lower = &t_old >> 2;
    if lower < BigUint::one() {
        lower = BigUint::one();
    }
    // upper = target_old * 4
    let upper_unclamped = &t_old << 2;
    let upper = core::cmp::min(upper_unclamped, pow_limit);

    if t_new < lower {
        t_new = lower;
    }
    if t_new > upper {
        t_new = upper;
    }

    biguint_to_bytes32(&t_new)
}

pub fn pow_check(header_bytes: &[u8], target: [u8; 32]) -> Result<(), TxError> {
    if header_bytes.len() != BLOCK_HEADER_BYTES {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "pow: invalid header length",
        ));
    }
    let target_bi = BigUint::from_bytes_be(&target);
    let pow_limit = BigUint::from_bytes_be(&POW_LIMIT);
    if target_bi.is_zero() || target_bi > pow_limit {
        return Err(TxError::new(
            ErrorCode::BlockErrTargetInvalid,
            "target out of range",
        ));
    }
    let h = block_hash(header_bytes)?;
    if h >= target {
        return Err(TxError::new(ErrorCode::BlockErrPowInvalid, "pow invalid"));
    }
    Ok(())
}

fn biguint_to_bytes32(x: &BigUint) -> Result<[u8; 32], TxError> {
    let b = x.to_bytes_be();
    if b.len() > 32 {
        return Err(TxError::new(ErrorCode::TxErrParse, "u256: overflow"));
    }
    let mut out = [0u8; 32];
    out[32 - b.len()..].copy_from_slice(&b);
    Ok(out)
}

File: clients/go/consensus/pow.go
package consensus

import (
	"bytes"
	"math/big"
)

// RetargetV1 computes the next target using the consensus formula in CANONICAL §15.
//
// All arithmetic is done with arbitrary precision; the result MUST fit in 32 bytes.
func RetargetV1(targetOld [32]byte, timestampFirst uint64, timestampLast uint64) ([32]byte, error) {
	var tActual uint64
	if timestampLast <= timestampFirst {
		tActual = 1
	} else {
		tActual = timestampLast - timestampFirst
	}
	return retargetV1WithActual(targetOld, tActual)
}

// RetargetV1Clamped computes retarget using clamped per-block timestamps (CANONICAL §15).
// The caller MUST provide exactly WINDOW_SIZE timestamps for the retarget window.
func RetargetV1Clamped(targetOld [32]byte, windowTimestamps []uint64) ([32]byte, error) {
	if len(windowTimestamps) != int(WINDOW_SIZE) {
		var zero [32]byte
		return zero, txerr(TX_ERR_PARSE, "retarget: invalid window timestamp count")
	}
	first := windowTimestamps[0]
	prev := first
	maxStep := uint64(MAX_TIMESTAMP_STEP_PER_BLOCK)

	for i := 1; i < len(windowTimestamps); i++ {
		lo, err := addU64(prev, 1)
		if err != nil {
			var zero [32]byte
			return zero, txerr(TX_ERR_PARSE, "retarget: timestamp clamp overflow")
		}
		hi, err := addU64(prev, maxStep)
		if err != nil {
			var zero [32]byte
			return zero, txerr(TX_ERR_PARSE, "retarget: timestamp clamp overflow")
		}
		v := windowTimestamps[i]
		if v < lo {
			v = lo
		} else if v > hi {
			v = hi
		}
		prev = v
	}

	tActual := prev - first
	if tActual == 0 {
		tActual = 1
	}
	return retargetV1WithActual(targetOld, tActual)
}

func retargetV1WithActual(targetOld [32]byte, tActual uint64) ([32]byte, error) {
	powLimit := new(big.Int).SetBytes(POW_LIMIT[:])
	tOld := new(big.Int).SetBytes(targetOld[:]) // big-endian
	if tOld.Sign() == 0 {
		var zero [32]byte
		return zero, txerr(TX_ERR_PARSE, "retarget: target_old is zero")
	}
	if tOld.Cmp(powLimit) > 0 {
		var zero [32]byte
		return zero, txerr(TX_ERR_PARSE, "retarget: target_old above pow_limit")
	}

	tExpected := uint64(TARGET_BLOCK_INTERVAL) * uint64(WINDOW_SIZE)
	if tExpected == 0 {
		var zero [32]byte
		return zero, txerr(TX_ERR_PARSE, "retarget: t_expected is zero")
	}

	// floor(target_old * T_actual / T_expected)
	num := new(big.Int).Mul(tOld, new(big.Int).SetUint64(tActual))
	den := new(big.Int).SetUint64(tExpected)
	tNew := new(big.Int).Div(num, den)

	// clamp lower = max(1, floor(target_old / 4))
	lower := new(big.Int).Rsh(new(big.Int).Set(tOld), 2) // /4
	if lower.Cmp(big.NewInt(1)) < 0 {
		lower.SetInt64(1)
	}
	// upper = target_old * 4
	upper := new(big.Int).Lsh(new(big.Int).Set(tOld), 2)
	if upper.Cmp(powLimit) > 0 {
		upper = powLimit
	}

	if tNew.Cmp(lower) < 0 {
		tNew = lower
	}
	if tNew.Cmp(upper) > 0 {
		tNew = upper
	}

	return bigIntToBytes32(tNew)
}

// PowCheck verifies integer(block_hash, be) < integer(target, be).
func PowCheck(headerBytes []byte, target [32]byte) error {
	targetInt := new(big.Int).SetBytes(target[:])
	powLimit := new(big.Int).SetBytes(POW_LIMIT[:])
	if targetInt.Sign() == 0 || targetInt.Cmp(powLimit) > 0 {
		return txerr(BLOCK_ERR_TARGET_INVALID, "target out of range")
	}

	h, err := BlockHash(headerBytes)
	if err != nil {
		return err
	}
	if bytes.Compare(h[:], target[:]) >= 0 {
		return txerr(BLOCK_ERR_POW_INVALID, "pow invalid")
	}
	return nil
}

func bigIntToBytes32(x *big.Int) ([32]byte, error) {
	var out [32]byte
	if x.Sign() < 0 {
		return out, txerr(TX_ERR_PARSE, "u256: negative")
	}
	b := x.Bytes() // big-endian without leading zeros
	if len(b) > 32 {
		return out, txerr(TX_ERR_PARSE, "u256: overflow")
	}
	copy(out[32-len(b):], b)
	return out, nil
}
